docs/design/interpreter/compiled-code-interop.md

```diff
@@ -0,0 +1,69 @@
+
+The interpreter operates on a separate stack space that it maintains. Every local variable, including argument registers reside on this space. When an interpreter method starts executing, it expects the arguments to be present one after the other at the location pointed to by the stack pointer. This means that for a call exiting the interpreter, with a certain signature, we need to call a thunk that receives at least the location of the parameters and the target address and that moves each argument from the interpreter stack to the corresponding reg/native stack location followed by a native call to the target address. Once the call returns it should move the return value from the regs/stack to the interpeter stack. If we need to pass to native/compiled code a code pointer that can be used to entry the interpreter, we need to create a thunk that moves all arguments from the native regs/stack to a separate memory location. This thunk should also have embedded a pointer that identifies the interpreter method that we need to execute. The thunk should over pass the memory location where the arguments have been copied together with the interpreter method to execute, so the interpreter entry code can set up a new interpreter frame and begin executing the method code.
+
+### Interop with compiled code on native architecture (ex. arm64)
+
+##### Interpreter Exit
+
+Let's assume the interpreter needs to either do a pinvoke call or a call to a method that is present in a AOT compiled image. At the callsite, in interpreter code, we will know exactly the signature of the call as well as native code pointer that needs to be called. We could have two different invocation paths: either a per signature specialized path or a generic path.
+
+  1. Specialized path
+  
+     A specialized path means that we will have a thunk specialized per signature that can be used to call any method with that signature. The basic set of signatures that are needed by an application are easy to compute. We will require a separate thunk for every pinvoke signature as well as for every signature of a method that is aot-compiled, because each aot-compiled method can end up being called from the interpreter. While these thunks could be emitted directly in assembly code, it makes more sense to compile them as IL wrappers and emit them together with the rest of the managed code when aot compiling an assembly. Each one of these wrappers will receive as arguments the native pointer of the method to call and the address on the interpreter stack where the arguments reside. The wrapper will be able to compute the base address of every single argument and load the argument value. It will then proceed with executing the native call, followed by writting back the result of the call to the interpreter stack in place of the arguments.
```

Comment:
As subsequently the suggestion is to use IL wrappers, I presume we are talking about IL signatures, rather than native signatures (which would be more shareable, but have other disadvantages I suppose) ?